File: pipeline_plugins/components/static/components/atoms/cc/update_set_service_status/v1_1.js
/**
 * Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
 * Edition) available.
 * Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

(function () {
    $.atoms.cc_update_set_service_status = [
        {
            tag_code: "set_select_method",
            type: "radio",
            attrs: {
                name: gettext("传参形式"),
                hookable: false,
                items: [
                    {value: "name", name: gettext("Set名称")},
                    {value: "id", name: gettext("Set ID")},
                    {value: "custom", name: gettext("自定义")},
                ],
                default: "name",
                validation: [
                    {
                        type: "required"
                    }
                ],
            },
            events: [
                {
                    source: "set_select_method",
                    type: "init",
                    action: function () {
                        this.emit_event(this.tagCode, "change", this.value)
                    }
                },
            ]
        },
        {
            tag_code: "set_attr_id",
            type: "input",
            attrs: {
                name: gettext("集群属性ID"),
                placeholder: gettext("用英文','分割，集群范围中填写的值会在此处填写的属性 ID 的值上进行过滤"),
                hookable: true,
                validation: [
                    {
                        type: "custom",
                        args: function (value) {
                            var result = {
                                result: true,
                                error_message: ""
                            };
                            var self = this;
                            if (!self.get_parent) {
                                return result
                            } else if (self.get_parent().get_child('set_select_method').value === "custom" && value.length===0 ) {
                                result.result = false;
                                result.error_message = gettext("传参形式为自定义时必填");
                            }

                            return result;
                        }
                    }
                ]
            },
            events: [
                {
                    source: "set_select_method",
                    type: "change",
                    action: function (value) {
                        var self = this;
                        if (value === "custom") {
                            self.show();
                        } else {
                            self._set_value('');
                            self.hide();
                        }
                    }
                }
            ],
        },
        {
            tag_code: "set_list",
            type: "textarea",
            attrs: {
                name: gettext("集群范围"),
                placeholder: gettext("多个集群使用英文','分割"),
                hookable: true,
                validation: [
                    {
                        type: "required"
                    }
                ]
            },
        },
        {
            tag_code: "set_status",
            type: "select",
            attrs: {
                name: gettext("服务状态"),
                allowCreate: true,
                placeholder: gettext('请选择'),
                hookable: true,
                remote: true,
                remote_url: $.context.get('site_url') + 'pipeline/cc_search_status_options/' + $.context.getBkBizId() + '/',
                remote_data_init: function (resp) {
                    if (resp.result === false) {
                        show_msg(resp.message, 'error');
                    }
                    return resp.data;
                },
                validation: [
                    {
                        type: "required"
                    }
                ]
            },
        },
    ]

})();

File: pipeline_plugins/components/collections/sites/open/cc/update_set_service_status/v1_1.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at
http://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""

import logging
from functools import partial

from django.utils.translation import gettext_lazy as _
from pipeline.component_framework.component import Component
from pipeline.core.flow.activity import Service
from pipeline.core.flow.io import StringItemSchema

from api.utils.request import batch_request
from gcloud.conf import settings
from gcloud.utils.handlers import handle_api_error
from pipeline_plugins.base.utils.inject import supplier_account_for_business
from packages.bkapi.bk_cmdb.shortcuts import get_client_by_username

logger = logging.getLogger("celery")

__group_name__ = _("配置平台(CMDB)")
VERSION = "1.1"

cc_handle_api_error = partial(handle_api_error, __group_name__)


class CCUpdateSetServiceStatusService(Service):
    def inputs_format(self):
        return [
            self.InputItem(
                name=_("传参形式"),
                key="set_select_method",
                type="string",
                schema=StringItemSchema(
                    description=_("集群填入方式，Set名称(name)，Set ID(id)，自定义（根据集群属性过滤）"),
                    enum=["name", "id", "custom"],
                ),
            ),
            self.InputItem(
                name=_("集群属性ID"),
                key="set_attr_id",
                type="string",
                schema=StringItemSchema(description=_("集群范围中填写的值会在此处填写的属性 ID 的值上进行过滤")),
            ),
            self.InputItem(
                name=_("集群范围"),
                key="set_list",
                type="string",
                schema=StringItemSchema(description=_("集群范围，多个集群使用英文','分割")),
            ),
            self.InputItem(
                name=_("服务状态"),
                key="set_status",
                type="string",
                schema=StringItemSchema(description=_("实时拉取的服务状态")),
            ),
        ]

    def outputs_format(self):
        return []

    def execute(self, data, parent_data):
        executor = parent_data.get_one_of_inputs("executor")
        tenant_id = parent_data.get_one_of_inputs("tenant_id")

        client = get_client_by_username(executor, stage=settings.BK_APIGW_STAGE_NAME)
        bk_biz_id = parent_data.get_one_of_inputs("bk_biz_id")
        supplier_account = supplier_account_for_business(bk_biz_id)
        set_list = data.get_one_of_inputs("set_list")
        cc_set_select = set_list.split(",")
        set_select_method = data.get_one_of_inputs("set_select_method")
        set_status = data.get_one_of_inputs("set_status")
        set_attr_id = data.get_one_of_inputs("set_attr_id") or "bk_set_name"

        bk_set_ids = []
        if set_select_method in ("name", "custom"):
            for set_name in cc_set_select:
                cc_search_set_kwargs = {
                    "bk_biz_id": int(bk_biz_id),
                    "fields": ["bk_set_id", set_attr_id],
                    "condition": {set_attr_id: set_name},
                }
                cc_search_set_result = batch_request(
                    client.api.search_set,
                    cc_search_set_kwargs,
                    path_params={"bk_supplier_account": supplier_account, "bk_biz_id": bk_biz_id},
                    headers={"X-Bk-Tenant-Id": tenant_id},
                )
                if not cc_search_set_result:
                    self.logger.error("batch_request client.cc.search_set error")
                    data.set_outputs("ex_data", "batch_request client.cc.search_set error")
                    return False
                for bk_set in cc_search_set_result:
                    if bk_set[set_attr_id] == set_name:
                        bk_set_ids.append(bk_set["bk_set_id"])
        elif set_select_method == "id":
            bk_set_ids = cc_set_select

        for set_id in bk_set_ids:
            cc_kwargs = {
                "bk_biz_id": bk_biz_id,
                "bk_supplier_account": supplier_account,
                "bk_set_id": set_id,
                "data": {"bk_service_status": set_status},
            }
            cc_result = client.api.update_set(
                cc_kwargs,
                path_params={"bk_biz_id": bk_biz_id, "bk_set_id": set_id},
                headers={"X-Bk-Tenant-Id": tenant_id},
            )
            if not cc_result["result"]:
                message = cc_handle_api_error("cc.update_set", cc_kwargs, cc_result)
                self.logger.error(message)
                data.set_outputs("ex_data", message)
                return False
        return True


class CCUpdateSetServiceStatusComponent(Component):
    """
    @version log （v1.1）:修改服务状态
    """

    name = _("修改集群服务状态")
    code = "cc_update_set_service_status"
    bound_service = CCUpdateSetServiceStatusService
    form = "{static_url}components/atoms/cc/update_set_service_status/v{ver}.js".format(
        static_url=settings.STATIC_URL, ver=VERSION.replace(".", "_")
    )
    version = VERSION
